runtime/src/global_request_scheduler_mts.c

```diff
@@ -0,0 +1,333 @@
+
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_if_earlier(struct sandbox_request **removed_sandbox_request,
+                                                       uint64_t                 target_deadline)
+{
+	/* This function won't be used with the MTS scheduler. Keeping merely for the polymorhism. */
+	return -1;
+}
+
+/**
+ * @param removed_sandbox_request pointer to set to removed sandbox request
+ * @param target_deadline the deadline that the request must be earlier than to dequeue
+ * @param mt_class the multi-tenancy class of the global request to compare the target deadline against
+ * @returns 0 if successful, -ENOENT if empty or if request isn't earlier than target_deadline
+ */
+int
+global_request_scheduler_mts_remove_with_mt_class(struct sandbox_request **removed_sandbox_request,
+                                                          uint64_t                 target_deadline,
+                                                          enum MULTI_TENANCY_CLASS target_mt_class)
+{
+	/* Avoid unnessary locks when the target_deadline is tighter than the head of the Global runqueue */
+	uint64_t global_guaranteed_deadline = priority_queue_peek(global_request_scheduler_mts_guaranteed);
+	uint64_t global_default_deadline    = priority_queue_peek(global_request_scheduler_mts_default);
+
+	switch (target_mt_class) {
+	case MT_GUARANTEED:
+		if (global_guaranteed_deadline >= target_deadline) return -ENOENT;
+		break;
+	case MT_DEFAULT:
+		if (global_guaranteed_deadline == UINT64_MAX && global_default_deadline >= target_deadline)
+			return -ENOENT;
+		break;
+	}
+
+	struct module_global_request_queue *top_mgrq          = NULL;
+	struct priority_queue *             destination_queue = global_request_scheduler_mts_guaranteed;
+
+	LOCK_LOCK(&global_lock);
+
+	/* Spot the Module Global Request Queue (MGRQ) to remove the sandbox request from */
+	int rc = priority_queue_top_nolock(destination_queue, (void **)&top_mgrq);
+	if (rc == -ENOENT) {
+		if (target_mt_class == MT_GUARANTEED) goto done;
+
+		destination_queue = global_request_scheduler_mts_default;
+
+		rc = priority_queue_top_nolock(destination_queue, (void **)&top_mgrq);
+		if (rc == -ENOENT) goto done;
+	} else {
+		if (top_mgrq->mt_class == MT_GUARANTEED && top_mgrq->module->remaining_budget <= 0) {
+			global_request_scheduler_mts_demote_nolock(top_mgrq);
+			debuglog("Demoted '%s' GLOBALLY", top_mgrq->module->name);
+			top_mgrq->mt_class = MT_DEFAULT;
+			rc                 = -ENOENT;
+			goto done;
+		}
+	}
+
+	assert(top_mgrq);
+
+	/* Remove the sandbox from the corresponding MGRQ */
+	rc = priority_queue_dequeue_nolock(top_mgrq->sandbox_requests, (void **)removed_sandbox_request);
+	if (rc == -ENOENT) goto done;
```

Comment:
not sure how this could happen. We hold the lock and found it be top before. assertion here?

Reply:
Nice catch! Noted!